### PR TITLE
Update truncate() documentation and testcases

### LINKED ISSUE #11382: truncate() raises IndexError on large timeseries
Hi,

I found a bug, tested on pandas 0.16 and 0.17.  On large time series, the ``truncate()`` method throws an IndexError.  Here's code that reproduces the bug:

```python
>>> import pandas as pd

>>> # Works fine on small timeseries:
>>> small = pd.Series(data=range(int(2e3)), index=range(int(2e3)))
>>> small.truncate()
>>> small.truncate(before=0, after=3e3)

>>> # IndexError thrown on large timeseries:
>>> big = pd.Series(data=range(int(2e6)), index=range(int(2e6)))
>>> big.truncate()
>>> big.truncate(before=0, after=3e6)
```

Info version: 

INSTALLED VERSIONS

commit: None
python: 2.7.8.final.0
python-bits: 64
OS: Linux
OS-release: 3.16.0-41-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8

pandas: 0.17.0
nose: 1.3.4
pip: 7.1.2
setuptools: 5.5.1
Cython: 0.20.2
numpy: 1.8.2
scipy: 0.14.0
statsmodels: None
IPython: 3.0.0
sphinx: 1.2.2
patsy: None
dateutil: 2.4.2
pytz: 2015.6
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.3.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.3.2
html5lib: None
httplib2: 0.9
apiclient: None
sqlalchemy: 0.9.7
pymysql: None
psycopg2: None

### PR DESCRIPTION
**Pull request by gliptak**
- [x] closes #11382 
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/test_panel.py [hunk 1]
```diff
@@ -1560,6 +1560,17 @@ def test_truncate_fillna_bug(self):
         # it works!
         result.fillna(value=0.0)
 
+    def test_truncate_outofbounds(self):
```
Comment (jreback) [inline pandas/tests/test_panel.py]: let's move the ``.truncate`` tests to ``test_generic`` instead.

Comment (gliptak) [inline pandas/tests/test_panel.py]: Done.

Comment (gliptak) [inline pandas/tests/test_panel.py]: Done

**Comment by gliptak** [comment]
I'm not sure what to make of those codecov failures ...

File: pandas/tests/test_panel.py (hunk 1 above)
Comment (jreback) [inline pandas/tests/test_panel.py]: and use the generic features to test Series/DataFrame/Panel

Comment (gliptak) [inline pandas/tests/test_panel.py]: Done

Comment (jreback) [inline pandas/tests/test_panel.py]: you said done, but it doesn't appear to be pushed. you need to use ``_construct``

Comment (gliptak) [inline pandas/tests/test_panel.py]: The latest (force) push uses ```assert_series_equal``` below. Were you pointing out something different you would like to see?

Comment (jreback) [inline pandas/tests/test_panel.py]: you need to structure the test so its runnable in all of the sub-classses Testers so its tested on ``Series/DataFrame`` etc.

Comment (gliptak) [inline pandas/tests/test_panel.py]: I got you. I ran into issues (memory, indeces) when trying to generalize, so decided to move it under Series (only).

**Comment by jreback** [comment]
just do the big selection across the 0th dimension

**Comment by gliptak** [comment]
The actual tests were OK (although I seen OutOfMemory exceptions ...), but the ```_construct``` wasn't.

For ```Series/DataFrame``` ```self._construct(,index=,...)``` vs for ```Panel/Panel4D``` ```self._construct(,major_axis=,minor_axis=,...)``` In this test I have to have the particular index.

How do you merge these?

**Comment by gliptak** [comment]
Are you OK with this being under ```Series```?

**Comment by jreback** [comment]
you need to make this generic, so no

**Comment by gliptak** [comment]
@jreback Please offer some snippets on how you see this being make generic (I described what I ran into in the previous post). Thanks

**Comment by gliptak** [comment]
@jreback How do you see this moving forward? Thanks

**Comment by jreback** [comment]
so in ``test_generic.py`` you can construct something generic like this:

```
shape = [int(1e6)] + ([1] * len(self._ndim-1)]
obj = self._construct(shape)
```

so will give you the first dim of 1e6 and the others 1

**Comment by gliptak** [comment]
@jreback Thank you. I made the changes, the 2e6 takes a long time to run locally (in Panel) ...

**Comment by gliptak** [comment]
Timing out in travis-ci too ```No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.```

**Comment by jreback** [comment]
thanks @gliptak 

the panel comparator was doing things a very slow way. all fixed up.

**Comment by gliptak** [comment]
@jreback Thank you for fixing Panel comparator.

**Pull request closed without merging**